piqueserver/config.py

```diff
@@ -0,0 +1,86 @@
+import toml
+from pprint import pprint
+
+class ConfigException(Exception):
+    pass
+
+
+class ConfigEngine():
+    '''
+    configuration engine that handles loading and saving config,
+    and all the things
+    '''
+    def __init__(self):
+        self.raw_config = {}
+        self.style = 'TOML'
+
+    def load_config(self, config_file, style='TOML'):
+        # TODO: could also have this merge into current config if available
+        self.style = style
+
+        if style == 'TOML':
+            self.raw_config = toml.load(open(config_file))
+        else:
+            raise Exception()
+
+    def dump_config(self, out_file=None, style='TOML'):
+        if out_file is None:
+            pprint(self.raw_config)
+        # TODO
+        pass
+
+    def get(self, name, default):
+        # TODO: handle nested config values
+        # XXX: should requesting a config value that is a dictionary object
+        #       return the whole dictionary, or fail?
+        return self.raw_config.get(name, default)
+
+    def set(self, name, value):
+        self.raw_config[name] = value
+
+    def __call__(self, name, cast=None, default=None, validate=None):
+        return ConfigObject(self, name, cast, default, validate)
+
+
+
+class ConfigObject():
+    '''
+    configuration object stored in the engine
+    '''
+    def __init__(self, engine, name, cast, default, validate):
+        self.name = name
+        self.cast = cast
+        self.default = default
+        self.validate = validate
+        self.engine = engine
+
+        value = self.engine.get(name, default)
```

Comment:
Interesting, this is the reverse of what I would have done. I would have kept a list of ConfigObjects in the Engine and then trigger the casting and validation from there if a value is set. We have to keep in mind that for anything except scripts, the configuration is not actually available at the time of creating this object yet.

Reply:
aaah yeah didn't think of that....